RFC-0011-InferenceMode.md

```diff
@@ -0,0 +1,138 @@
+Note: a large part of this RFC will become "InferenceMode" documentation once it's finalized.
+
+## Goals:
+- Provide a RAII in C++ and a context manager in Python frontend to switch between inference mode and normal mode, with the following constraints:
+  - correctness is always guaranteed. (compared to `AutoNonVariableType` which has risks producing silent wrong result.)
+  - performance of inference mode should match current existing `AutoNonVariableTypeMode` which is widely used in prod.
+  - switching between normal mode and inference mode should be really easy with minimal code change.
+- Make `AutoNonVariableTypeMode` an internal only API, replace all callsites of `AutoNonVariableTypeMode` outside pytorch codebase with the new `InferenceMode`.
+
+## Non-goals:
+- Match the theoretical best inference performance which can be achieved by stripping all autograd related stuff at build time (not flexible).
```

Comment:
Actually, are we sure about this? If you write code solely in InferenceMode, with no interaction with non-InferenceMode tensors, it seems to me that theoretical best performance should be attainable (since we never have to hit the safety code for the intermixing situation).

Reply:
oh yes, without the intermixing situation (and maybe some startup cost initializing dispatch table) the performance should be attainable. :D